Add insecure repository support

fahhem opened the pull request:
Local registries, such as microk8s, don't have SSL setup.

This PR just adds the ability to specify to push to an insecure repository. I'm not sure if that's the intended design pattern for this, so I'm open to suggestions!

fahhem committed:
Add insecure repository support (for microk8s)

fahhem commented:
/assign @smukherj1

fahhem committed:
Buildifier - unsorted-dict-items

fahhem commented:
This might also address #1245

smukherj1 reviewed (commented):
Thanks for the PR! Do you mind adding a test [here](https://github.com/bazelbuild/rules_docker/blob/master/testing/e2e/pusher.sh) that exercises the new insecure_repository attribute?

jake-arkinstall commented:
Any update on this?

asv commented:
@fahhem will you finish this PR or can I post mine? For our production, this is a very important feature in rules_docker.

fahhem commented:
I don't have a publicly-available insecure repository available to add a test against. If you have a better PR, go ahead and post it, I'm not attached to this PR in any way. I was simply sharing what I had to do to get a on-machine docker repository to work.

fahhem committed:
add test

fahhem commented:
I've rebased onto master. I don't know the test infra here, so I added a "test" that is probably wrong. :shrug: if anyone else has a better test or at least a suggestion on how to write one, I'm happy to swap mine out for it.

cfife-btig commented:
We are very interested in having this merged in. We are also dependent on a registry which is insecure and are having to write lots of workarounds without this change.

fahhem committed:
Merge branch 'master' into add_insecurity

fahhem commented:
Rebased again. This time the tests are failing for the same reason the master tests are failing (something about an unsupported version of ubuntu?  `ubuntu2104` vs the mentioned `ubuntu1804`)

fahhem commented:
@uhthomas any chance this can be reviewed? It's a pretty small PR :)

fahhem committed:
Merge branch 'bazelbuild:master' into add_insecurity

linzhp reviewed (approved):
LGTM

fahhem commented:
@linzhp any idea on what I should do about the required-but-failing CI step? Or will you be able to merge it anyway?

fahhem commented:
Oh, and thank you!

linzhp commented:
It doesn't allow me to merge without CI passing

uhthomas commented on container/go/cmd/pusher/pusher.go:
```diff
@@ -126,7 +127,12 @@ func main() {
 	}
 
-	if err := push(stamped, img); err != nil {
+	var options = []name.Option{};
```
This needs to be run through gofmt, Go code does not typically have semi colons.

Also, it's usually preferred to write this as:

```go
var opts []name.Option
```

uhthomas commented on container/go/cmd/pusher/pusher.go:
```diff
@@ -126,7 +127,12 @@ func main() {
-	if err := push(stamped, img); err != nil {
+	var options = []name.Option{};
+	if *insecureRepository {
+		options = append(options, name.Insecure);
```
gofmt

cfife-btig commented:
I'm just happy to see some progress is on this MR :')

fahhem committed:
gofmt

fahhem commented:
Same here :D Thanks @uhthomas, I ran gofmt and it had a couple more changes than around my code, but it was all whitespace stuff.

fahhem committed:
fix rename

fahhem commented:
sorry for the noise, github's reviewer UI looks... strange. Anyone consider reviewable.io?

fahhem commented:
@uhthomas sorry to bother!

cfife-btig commented:
I don't mean to keep pecking either. But my team is also really interested in getting this merged in as well. I know that the CI is still being blocked but it sorta sounds like it just might be some sort of noise and not anything legitimate. Although of course I also have no visibility.

uhthomas commented:
Looks okay, we'll need to understand why CI isn't happy and then we can merge.

fahhem commented:
All PR's are blocked until CI is fixed? Is anyone at bazel working on fixing the CI?

It's a lot easier to fix buildkite stuff when you have access to the servers, but I can take a stab at it still.

linzhp commented:
Sorry for the delay. I just fixed CI (#2178)

fahhem committed:
Trigger CI again

fahhem committed:
Undo change

fahhem committed:
Merge branch 'bazelbuild:master' into add_insecurity

fahhem committed:
Fix docs

fahhem committed:
update docs properly

fahhem commented:
Thank you @linzhp I fixed the tests too, so this PR should be ready to go!

cfife-btig commented:
🎉🎉🎉

yc185050 commented:
Any idea when will this be released? Waiting on this fix.

fahhem commented:
@yc185050 I recommend opening a new issue to request a release. Personally, I'm using master at that merged commit as my company's release while we wait for an official release

matthewei commented:
Hi , how can we use this feature?
